### ISSUE TITLE
Ubuntu 14.04 LTS and build_deb.sh problem

### ISSUE THREAD

**Issue by bitinn** [shadowsocks/shadowsocks-libev#1461]
### What version of shadowsocks-libev are you using?

Latest source

### What operating system are you using?

Ubuntu 14.04 LTS

### What did you do?

Following the build step using `build_deb.sh`.

### What did you expect to see?

Binary files in `build-area` folder or at `/usr/local/bin`

### What did you see instead?

```
dpkg: error processing archive simple-obfs_*.deb (--install):
 cannot access archive: No such file or directory
Errors were encountered while processing:
 simple-obfs_*.deb
Reading package lists... Done
Building dependency tree       
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 3 not upgraded.
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package libbloom-dev
E: Unable to locate package libmbedtls-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package libbloom-build-deps
E: Unable to locate package mbedtls-build-deps
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package shadowsocks-libev-build-deps
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package golang-github-klauspost-reedsolomon-build-deps
E: Unable to locate package golang-github-xtaci-kcp-build-deps
E: Unable to locate package golang-github-xtaci-smux-build-deps
E: Unable to locate package kcptun-build-deps
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package golang-github-urfave-cli-build-deps
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package golang-github-golang-snappy-build-deps
E: Unable to locate package dh-golang-build-deps
E: Unable to locate package golang-github-pkg-errors-build-deps
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package golang-github-klauspost-reedsolomon-dev
E: Unable to locate package golang-github-xtaci-kcp-dev
E: Unable to locate package golang-github-xtaci-smux-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package golang-github-urfave-cli-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package golang-github-pkg-errors-dev
E: Unable to locate package golang-github-golang-snappy-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 3 not upgraded.
```

After which the script exits without throwing errors.

### What is your config in detail (with all sensitive info masked)?

There is not additional config.

------

As a side question, I never did find out the reason for abandoning PPA and Self-hosted Archive?

Most people just want a working copy of `ss-server` in `/usr/local/bin`, it doesn't need to be from latest source or have latest features.

------

EDIT:

Tried the generic Linux build step and encounter this problem:

```
make[2]: Entering directory `/home/dev/shadowsocks-libev/src'
  CCLD     ss-local
ss_local-aead.o: In function `aead_cipher_decrypt':
/home/dev/shadowsocks-libev/src/aead.c:246: undefined reference to `crypto_aead_xchacha20poly1305_ietf_decrypt'
ss_local-aead.o: In function `aead_cipher_encrypt':
/home/dev/shadowsocks-libev/src/aead.c:207: undefined reference to `crypto_aead_xchacha20poly1305_ietf_encrypt'
collect2: error: ld returned 1 exit status
make[2]: *** [ss-local] Error 1
make[2]: Leaving directory `/home/dev/shadowsocks-libev/src'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/home/dev/shadowsocks-libev'
make: *** [all] Error 2
```

I find this strange as:

- I did build libsodium from source (v1.0.11).
- running `sudo apt-get purge libsodium-dev`, then rebuild libsodium allow the build to pass.

**Comment by wongsyrone** [comment]
> running sudo apt-get purge libsodium-dev, then rebuild libsodium allow the build to pass.

It indicates old headers get included by cpp.

**Comment by bitinn** [comment]
@wongsyrone Thx for the suggestion, as I didn't install libsodium previously, so I am confused how that could happen.

------

Anyway, just notice building from source will definitely not setup default config or init.d script. I wonder if we can somehow setup a backported PPA (even if with out-of-date build).

I wonder the rational behind that, maybe people were being negative when asking for updates, maybe Ubuntu just isn't getting much love these days. Either way, I would like to say there are many people aren't capable or comfortable building from source, so some automated solutions would be great (maybe an official docker guide if PPA is too much).

**Comment by madeye** [comment]
@bitinn Try docker: https://github.com/shadowsocks/shadowsocks-libev/tree/master/docker/alpine

**Comment by bitinn** [comment]
Thx. I opted to the generic Linux build route as its cleaner (or more to my taste).

On a side note, would you guys be ok if I write a simple wiki page for Generic Build setup steps and link to it on readme? To get to similar working state on Ubuntu you at least need to setup a process monitor and a config file. Which you don't have when doing generic build.

(From my phone)

**Comment by madeye** [comment]
> Generic Build setup steps

https://github.com/shadowsocks/shadowsocks-libev#linux

**Comment by bitinn** [comment]
@madeye Nah, I am not talking about that guide.

I mean after that guide, one would need to do this: https://gist.github.com/euyuil/55cfe14e6b3ba9ce416e

Almost too simple to include, I know. But judging from the number of blog posts/gists about "how to setup Shadowsocks on Ubuntu version X", we have fixed that problem with a simple update on Readme or Wiki page.

**Comment by madeye** [comment]
@bitinn The only recommended way to install shadowsocks-libev is through `apt install shadowsocks-libev` on the latest Ubuntu/Debian distributions.

We've no plan to introduce any official "one-key-install" script.

**Comment by bitinn** [comment]
@madeye It seems there is a misunderstanding.

At this point I am not arguing you should introduce anything that automate the process. It would be nice for 14.04/16.04 LTS users, but if you don't have the resource, I am not going to push it.

What I am saying is: given 14.04/16.04 LTS users mostly install from source. It would be nice to provide the complete setup guide. Current guide only goes to building binary. But I am pretty sure `apt install shadowsocks-libev` does more than that, it setups the `init.d` script as well as the config folder `/etc/shadowsocks-libev`. It would be nice to include those simple steps in Readme.

**Comment by madeye** [comment]
@bitinn For Debian/Ubuntu package building, please refer to https://github.com/shadowsocks/shadowsocks-libev#build-deb-package-from-source

**Comment by hosiet** [comment]
I think your need seems to be installing init.d script and systemd service files when doing `make install` instead of letting distribution packagers copy them explicitly.

That's doable, but we will need some tweaks.

@madeye, will you accept such PR to deal with this problem?

**Comment by madeye** [comment]
@hosiet Nope, I choose to keep all distribution dependent files from `make install` target. In other words, 'Makefile` should not include any distribution dependent rules (only platform / arch related rules).

**Comment by bitinn** [comment]
I am closing this as it's not going anywhere.

My original problem is about `build_deb.sh` not building properly, and it's the same problem that should be fixed with #1448

**Comment by dc0d** [comment]
Same error - Ubuntu 16.04, 2018 05 05.
